core/vm/evm.go

```diff
@@ -838,7 +830,54 @@ func (evm *EVM) captureOperationMode(isWriteOperation bool) error {
 	return nil
 }
 
-// enforceMultitenancyCheck returns true if EVM is enforced to do multitenancy check, false otherwise
+// Quorum
+//
+// captureAffectedContract stores the contract address to the affectedContract list if not yet there.
+// The affected mode is also updated if required.
+// Default affected reason is MessageCall.
+func (evm *EVM) captureAffectedContract(address common.Address, mode AffectedMode) error {
+	if _, ok := evm.affectedContracts[address]; !ok {
+		evm.affectedContracts[address] = newAffectedType(MessageCall, mode)
+	}
+	return evm.updateAffectedMode(address)
+}
+
+// Quorum
+//
+// updateAffectedMode is usied in simulation/eth_call for multitenancy, it sets the expectation of AffectedMode
+// to be verified later when an opcode is executed.
+func (evm *EVM) updateAffectedMode(address common.Address) error {
```

Comment:
`updateAffectedMode` looks to do a lot could we instead do: 

1. ```(mode AffectedMode) Update(authorizedRead, authorizedWrite bool) AffectedMode``` that updates the mode
2. Update ```(evm *EVM) captureAffectedContract(...) *AffectedType``` (returning the Affected type but not doing `updateAffectedMode` so not returning any error)
3. In `getDualState` do
```
affectedType, err := evm.captureAffectedMode(...)
...
if evm.AuthorizeMessageCallFunc != nil {
    authorizedRead, authorizedWrite, err := evm.AuthorizeMessageCallFunc(address)
	...
	affectedType.mode.Update(authorizedRead, authorizedWrite)
}
```